README.md
# VOSM

## What's VOSM for?

Vision Open Statistical Model ([VOSM](https://github.com/jiapei100/VOSM)) contains variants of 2D Statistical Models, namely, variants of [ASMs](https://en.wikipedia.org/wiki/Active_shape_model) and [AAMs](https://en.wikipedia.org/wiki/Active_appearance_model). It's mainly composed of two parts:

* Building: build the statistical models (both 2D and 3D are supported in [VOSM](https://github.com/jiapei100/VOSM)). For now, the following models can be built by specifying different parameters:
  - Shape Model: ```-t "SM"```, fundamental shape model
  - Texture Model: ```-t "TM"```, fundamental texture model
  - [Appearance Model](http://www.imm.dtu.dk/~aam/main/node15.html): ```-t "AM"```, fundamental appearance model, originated from [Basic AAM](http://www.imm.dtu.dk/~aam/main/node16.html)
  - Image Alignment appearance Model: ```-t IA```, inverse compositional image alignment, for [ICIA AAM](https://www.ri.cmu.edu/pub_files/pub4/matthews_iain_2004_2/matthews_iain_2004_2.pdf)
  - Feature Model: ```-t "FM"```, a generalized model based on **ANY** type of features adopted by the users
  - Shape model with Local Texture Constraints: ```-t "SMLTC"```, originated from [CLM](https://personalpages.manchester.ac.uk/staff/timothy.f.cootes/Models/clm.html)
  - SMNDPROFILE. ```-t "SMNDPROFILE"```

* Fitting: test the effects of fitting (**ONLY** 2D is suppprted in [VOSM](https://github.com/jiapei100/VOSM))


## Prerequisites

- [OpenCV](https://github.com/opencv/opencv): >= [Version 3.4.1](https://github.com/opencv/opencv/releases/tag/3.4.1).
- [Boost](https://www.boost.org/): >= [Version 1.65.1](https://www.boost.org/users/history/version_1_65_1.html).


## Download, Compile, and Build

```
$ git clone https://github.com/jiapei100/VOSM
$ mkdir build
$ cd build
$ ccmake ../
$ make -j8
$ sudo make install
```

Two commands and eight libraries will be built:
By default, 2 commands will be built under: **/usr/local/bin**:
- testsmbuilding
- testsmfitting

By default, 8 libraries will be built under: **/usr/local/lib**:
- libvosm_comalgs.so
- libvosm_common.so
- libvosm_cvcommon.so
- libvosm_ensembletraining.so
- libvosm_featureextraction.so
- libvosm_integraltransform.so
- libvosm_smbuilding.so
- libvosm_smfitting.so


## How to use VOSM?

Let's check the **help** info about both commands **testsmbuilding** amnd **testsmfitting**.

** testsmbulding**

```
$ testsmbuilding --help
Usage: test_smbuilding [options] save_directory annotation_directory image_directory shapeinfo_path database channels type levels percentage
options:
   -o    output directory (default './')
   -a    annotation directory (required)
   -i    image directory (required)
   -s    path of the file shapeinfo (required)
   -d    training database (required, IMM, AGING, BIOID, XM2VTS, FRANCK, EMOUNT, JIAPEI )
   -c    channels (1 or 3, default 3)
   -t    statistical model type (SM, TM, AM, IA, FM, SMLTC, SMNDPROFILE. default SMNDPROFILE )
   -l    level of parymid (default 4)
   -p    percentage of shape, texture and appearance PCA (default 0.95)
Note: If you are building SMLTC or SMNDPROFILE, you must specify
 -c 1, namely, SMLTC and SMNDPROFILE can only deal with gray-level images.
 ```

** testsmfitting**

```
$ testsmfitting --help
Usage: smfitting [options] trained_data type testing_images testing_annotations database staticORdynamic recording
options:
   -o   trained data DIRECTory (required)
   -t   fitting method to be used (ASM_PROFILEND, ASM_LTC, AAM_BASIC, AAM_CMUICIA, AAM_IAIA. default ASM_PROFILEND )
   -i   testing image DIRECTory containing at least one image (required)
   -a   testing annotation DIRECTory (can be ignored)
   -d   testing database -- if annotation DIRECTory is specified, database should also be specified for further evaluation on fitting performance (can be ignored)
   -s   static image sequence or dynamic image sequence (default value true)
   -r   recording the fitting results or not (default value false)


Note: current smfitting doesn't defaultly afford 1D Profile ASM.
If you would like to try 1D Profile ASM for static images in current smfitting,
you have to manually change the code in function VO_Fitting2DSM::VO_StartFitting
in file VO_Fitting2DSM.cpp, say, around line 290 of file VO_Fitting2DSM.cpp
change the 5th parameter from '2' to '1' of function
dynamic_cast<VO_FittingASMNDProfiles*>(this)->VO_ASMNDProfileFitting
If you would like to try 1D Profile ASM for dynamic image sequence in current smfitting,
you have to manually change the code in function
in file smfitting.cpp, say around line 453 of file smfitting.cpp
change the 5th parameter from '2' to '1' of function
dynamic_cast<VO_FittingASMNDProfiles*>(fitting2dsm)->VO_ASMNDProfileFitting


Face Detection: current smfitting use Adaboost technology to detect face as well as face components
for face location initialization. Refer to CFaceDetectionAlgs in main().
Default Adaboost detectors installed with OpenCV installation are used in current smfitting
You may manually change the Adaboost detectors according to your own cascade file paths


Face Tracking: current smfitting only deal with image sequences.
If static images are to be tested, there is no point to carry out tracking because
for every image, Adaboost detection will be carried out.
If dynamic image sequences are to be tested, current smfitting only affords Camshift tracking strategy.
Please Refer to CTrackingAlgs() in main(), the default setting of function CTrackingAlgs() is Camshift algorithm


Vision Open doesn't afford the video IO or webcam IO although JIA Pei has done his own IO for all kinds.
Users are highly encouraged to use their own video file IO and webcam IO and use VOSM in their own real-time applications.
```




## Key Relative Publications

1) P. JIA, [2D Statistical Models](http://www.visionopen.com/members/jiapei/publications/pei_sm2dreport2010.pdf), Technical Report of Vision Open Working Group, 2st Edition, Oct 21, 2010.

2) P. JIA. [Audio-visual based HMI for an Intelligent Wheelchair](http://www.visionopen.com/members/jiapei/publications/pei_thesischapter34.pdf). PhD thesis, University of Essex, 2010.

3) T. Cootes and C. Taylor. [Statistical models of appearance for computer
vision](http://www.face-rec.org/algorithms/aam/app_models.pdf). Technical report, Imaging Science and Biomedical Engineering, University of Manchester, March 8 2004.

4) I. Matthews and S. Baker. [Active appearance models revisited](https://www.ri.cmu.edu/pub_files/pub4/matthews_iain_2004_2/matthews_iain_2004_2.pdf). International Journal of Computer Vision, 60(2):135–164, November 2004.

5) M. B. Stegmann, [Active Appearance Models: Theory, Extensions and Cases](http://www2.imm.dtu.dk/~aam/main/), 2000.


CHANGES
Release notes of VOSM.

--------------------------- Release 0.3.4 --April, 2014--------------------
Coding style modification in this version VOSM 0.3.4.

1 General
   - VOSM 0.3.4 is based upon:
     
      a) Boost (>= 1.55.0) for file and directory operations
      b) OpenCV (2.4.8)

   - COPYRIGHT file has been modified a bit to be viewed more clearly.

   - "using namespace XXX" has been removed from this version VOSM 0.3.4,
   which is now using namespace in the style "std::vector vec;", etc.

--------------------------- Release 0.3.2 --December, 2010--------------------
Minor improvements are made in this version VOSM 0.3.2,
in order to be compatible with OpenCV 2.2.0 .

1 General
   - VOSM 0.3.2 is based upon:
     
      a) Boost (>= 1.34.1) for file and directory operations
      b) OpenCV (2.2.0)

   - INSTALL file has been changed to clarify how to install VOSM 0.3.2 .

   - HOWTOUSE file is added to clarify how to actually use VOSM 
	 by 2 key command lines .


--------------------------- Release 0.3.1 --November, 2010--------------------
Right after VOSM 0.3.0 was released on November 8, 2010,
due to prompt feedbacks from some VOSM users,
VOSM 0.3.1 is now released with the following improvements.

1 General
   - VOSM 0.3.1 is neither based on libgsl nor yaml-cpp any longer.
     The only 3rd party libraries that VOSM 0.3.1 is based upon are
     
      a) Boost (>= 1.34.1) for file and directory operations
      b) OpenCV (subervsion R3594 or above) as the basic but primary libraries 
         that VOSM 0.3.1 is built upon

   - A formal documentation is produced for reference

   - Add brief descriptions for every class


----------------------------- Release 0.3 --November, 2010--------------------

It has been a long time for Vision Open to release its recent 
research fruits related to ASM/AAMs.
VOSM 0.3 offers a thoroughly fresh coding implementations, 
and lots of improvements and new functionalities : 

1 General

  - Current VOSM can only be legally used under LGPL, namely,
	Lesser General Public License.

  - Current VOSM libraries requires CMake for configuration.

  - Changes in the set of supported platforms:

    - Fully supported by GNU g++ 4.5 and tested.

    - Well supported by Microsoft Windows Visual Studio 
	  C++ compiler 2008 (VC9). You may have to re-configure accordingly.

  - Changes in the set of supported OpenCV:

    - Must be OpenCV SVN R3594 or above 
	  (OpenCV 2.1 is too old to be compile with current VOSM 0.3)

2 Algorithms

  - Current VOSM affords 2D Profile ASM, which is dramatically faster
    than traditional 1D Profile ASM.

  - Current VOSM proposes a conceptual novel idea named as
    Active Feature Model.

  - Current VOSM implements all popular ASMs/AAMs,
    both for static images and dynamic videos, including:

	- 1D profile ASM

    - 2D profile ASM

    - Direct Local texture contrained (LTC) ASM

    - Basic AAM

    - ICIA(Inverse compositional image alignment) AAM

    - IAIA(Inverse additive image alignment) AAM

----------------------------- Release 0.2 -- April, 2008----------------------

Second release of "AAMBuilding":

    - No source, but a xwWidget based GUI application 
	  with some binary static libraries for model building.

----------------------------- Release 0.1 -- February, 2007-------------------

Initial release of "AAMBuilding", 

  	- Contains only the model building code. 

------------------------------------------------------------------------------




modules/smbuilding/src/VO_AXM.cpp
/****************************************************************************************************
*                                                                                                   *
*                   IMPORTANT: READ BEFORE DOWNLOADING, COPYING, INSTALLING OR USING.               *
*                                                                                                   *
*   By downloading, copying, installing or using the software you agree to this license.            *
*   If you do not agree to this license, do not download, install, copy or use the software.        *
*                                                                                                   *
*                                   License Agreement                                               *
*                           For Vision Open Statistical Models                                      *
*                                                                                                   *
* Copyright (C):    2006~2014 by JIA Pei, all rights reserved.                                      *
*                                                                                                   *
*                   VOSM is free software under the terms of the GNU Lesser General Public License  *
*                   (GNU LGPL) as published by the Free Software Foundation; either version 3.0 of  *
*                   the License, or (at your option) any later version.                             *
*                   You can use it, modify it, redistribute it, etc; and redistribution and use in  *
*                   source and binary forms, with or without modification, are permitted provided   *
*                   that the following conditions are met:                                          *
*                                                                                                   *
*                   a) Redistribution's of source code must retain this whole paragraph of          *
*                   copyright notice, including this list of conditions and all the following       *
*                   contents in this copyright paragraph.                                           *
*                                                                                                   *
*                   b) Redistribution's in binary form must reproduce this whole paragraph of       *
*                   copyright notice, including this list of conditions and all the following       *
*                   contents in this copyright paragraph, and/or other materials provided with      *
*                   the distribution.                                                               *
*                                                                                                   *
*                   c) The name of the copyright holders may not be used to endorse or promote      *
*                   products derived from this software without specific prior written permission.  *
*                                                                                                   *
*                   Any publications based on this code must cite the following five papers,        *
*                   technic reports and on-line materials.                                          *
*                                                                                                   *
*                   1) P. JIA, 2D Statistical Models, Technical Report of Vision Open Working       *
*                   Group, 2st Edition, Oct 21, 2010.                                               *
*                   http://www.visionopen.com/members/jiapei/publications/pei_sm2dreport2010.pdf    *
*                                                                                                   *
*                   2) P. JIA. Audio-visual based HMI for an Intelligent Wheelchair.                *
*                   PhD thesis, University of Essex, 2010.                                          *
*                   http://www.visionopen.com/members/jiapei/publications/pei_thesischapter34.pdf   *
*                                                                                                   *
*                   3) T. Cootes and C. Taylor. Statistical models of appearance for computer       *
*                   vision. Technical report, Imaging Science and Biomedical Engineering,           *
*                   University of Manchester, March 8 2004.                                         *
*                                                                                                   *
*                   4) I. cv::Matthews and S. Baker. Active appearance models revisited.                *
*                   International Journal of Computer Vision, 60(2):135–164, November 2004.         *
*                                                                                                   *
*                   5) M. B. Stegmann, Active Appearance Models: Theory, Extensions and Cases,      *
*                   http://www2.imm.dtu.dk/~aam/main/, 2000.                                        *
*                                                                                                   *
*                                                                                                   *
* Version:          0.3.4                                                                           *
* Author:           JIA Pei                                                                         *
* Contact:          jp4work@gmail.com                                                               *
* URL:              http://www.visionopen.com                                                       *
* Create Date:      2010-11-04                                                                      *
* Modify Date:      2014-04-17                                                                      *
****************************************************************************************************/


#include <fstream>
#include <sstream>
#include <string>

#include <boost/filesystem.hpp>

#include "VO_AXM.h"


VO_AXM::VO_AXM(unsigned int method, unsigned int levels)
{
	this->init(method, levels);
}


void VO_AXM::init(unsigned int method, unsigned int levels)
{
	this->m_iMethod					=	method;
    this->m_iNbOfPyramidLevels 		= 	levels;
}


VO_AXM::~VO_AXM()
{
	
}


/**
 * @author     	JIA Pei
 * @version    	2010-02-13
 * @brief      	Save ASM to a specified folder
 * @param      	fd         	Input - the folder that ASM to be saved to
 * @return		void
*/
void VO_AXM::VO_Save ( const std::string& fd )
{
	switch(this->m_iMethod)
	{
		case AAM_BASIC:
		case AAM_DIRECT:
		case AAM_FAIA:
		case AAM_CMUICIA:
		case AAM_IAIA:
		VO_TextureModel::VO_Save(fd);
		break;
		case ASM_PROFILEND:
		case ASM_LTC:
		VO_ShapeModel::VO_Save(fd);
		break;
		case CLM:
		case AFM:
		break;
	}
	
    // create AXM subfolder for just AXM model data
    std::string fn = fd+"/AXM";
    if (!boost::filesystem::is_directory(fn) )
        boost::filesystem::create_directory( fn );

    std::ofstream fp;
    std::string tempfn;

    // AXM
    tempfn = fn + "/AXM" + ".txt";
    fp.open(tempfn.c_str (), ios::out);
    fp << "m_iNbOfPyramidLevels" << endl << this->m_iNbOfPyramidLevels << endl;
    fp.close();fp.clear();
}


/**
 * @author     	JIA Pei
 * @version    	2010-02-13
 * @brief      	Load all trained data
 * @param      	fd         	Input - the folder that ASM to be saved to
 * @return		void
*/
void VO_AXM::VO_Load ( const std::string& fd )
{
	switch(this->m_iMethod)
	{
		case AAM_BASIC:
		case AAM_DIRECT:
		case AAM_FAIA:
		case AAM_CMUICIA:
		case AAM_IAIA:
		VO_TextureModel::VO_Load(fd);
		break;
		case ASM_PROFILEND:
		case ASM_LTC:
		VO_ShapeModel::VO_Load(fd);
		break;
		case CLM:
		case AFM:
		break;
	}
	
	std::string fn = fd+"/AXM";
    if (!boost::filesystem::is_directory(fn) )
    {
        cout << "AXM subfolder is not existing. " << endl;
        exit(EXIT_FAILURE);
    }

    std::ifstream fp;
    std::string tempfn;
    std::string temp;

    // AXM
    tempfn = fn + "/AXM" + ".txt";
    fp.open(tempfn.c_str (), ios::in);
    fp >> temp >> this->m_iNbOfPyramidLevels;					// m_iNbOfPyramidLevels
    fp.close();fp.clear();
}


/**
 * @author     	JIA Pei
 * @version    	2010-02-13
 * @brief      	Load all trained data for fitting
 * @param      	fd         	Input - the folder that ASM to be saved to
 * @return		void
*/
void VO_AXM::VO_LoadParameters4Fitting ( const std::string& fd )
{
	switch(this->m_iMethod)
	{
		case AAM_BASIC:
		case AAM_DIRECT:
		case AAM_FAIA:
		case AAM_CMUICIA:
		case AAM_IAIA:
		VO_TextureModel::VO_LoadParameters4Fitting(fd);
		break;
		case ASM_PROFILEND:
		case ASM_LTC:
		VO_ShapeModel::VO_LoadParameters4Fitting(fd);
		break;
		case CLM:
		case AFM:
		break;
	}
	
	std::string fn = fd+"/AXM";
    if (!boost::filesystem::is_directory(fn) )
    {
        cout << "AXM subfolder is not existing. " << endl;
        exit(EXIT_FAILURE);
    }

    std::ifstream fp;
    std::string tempfn;
    std::string temp;

    // AXM
    tempfn = fn + "/AXM" + ".txt";
    fp.open(tempfn.c_str (), ios::in);
    fp >> temp >> this->m_iNbOfPyramidLevels;					// m_iNbOfPyramidLevels
    fp.close();fp.clear();
}

